### PR TITLE
[WIP] Critic LSTM memories

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

The Actor/Critic split will not allow us to evaluate the critic memories during inference (i.e, in `policy.get_action()`. This PR introduces a workaround, where the values are computed during `get_trajectory_value_estimates()` in the optimizer. We have to compute the value and resulting memories per sequence in each trajectory that is passed to this function. 

Furthermore, initial memories per sequence were the current memories, not the previous (has been the case since forever). For instance, if you take a step at time t with obs_t, and get action_t and memory_t, when initializing the LSTM you want to use memory_(t-1). This PR fixes that for the policy memories. 

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: .yamato/com.unity.ml-agents-performance.yml [hunk 1]
```diff
@@ -1,6 +1,5 @@
 test_editors:
   - version: 2019.4
-  - version: 2020.1
```
Comment (andrewcoh) [inline .yamato/com.unity.ml-agents-performance.yml:3]: intentional?

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 2]
```diff
@@ -54,23 +55,71 @@ def create_reward_signals(self, reward_signal_configs):
...
+            _mem = memory
+            for seq_num in range(
+                resequenced_buffer.num_experiences // self.policy.sequence_length - 1
+            ):
+                seq_obs = []
+                for _obs in reseq_obs:
+                    start = seq_num * self.policy.sequence_length
+                    end = (seq_num + 1) * self.policy.sequence_length
+                    seq_obs.append(_obs[start:end])
+                _, next_seq_mem = self.critic.critic_pass(
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:103]: we can save the calls to get value estimates  later if we use the returned value here

Comment (ervteng) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:103]: +1 it is a TODO

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 3]
```diff
@@ -53,23 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
...
+        self, tensor_obs: List[torch.Tensor], initial_memory: np.ndarray
+    ) -> Tuple[Dict[str, torch.Tensor], AgentBufferField, torch.Tensor]:
+        """
+        Evaluate the batch sequence-by-sequence, assembling the result. This enables us to get the
+        intermediate memories for the critic.
+        """
+        num_experiences = tensor_obs[0].shape[0]
+        all_next_memories = AgentBufferField()
+        # The 1st sequence are the ones that are padded. So if seq_len = 3 and
+        # trajectory is of length 10, the ist sequence is [pad,pad,obs].
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:68]:
```suggestion
        # trajectory is of length 10, the 1st sequence is [pad,pad,obs].
```

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 4]
```diff
@@ -53,23 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
...
+            memory = self.critic_memory_dict[agent_id]
+        else:
+            memory = (
+                torch.zeros((1, 1, self.critic.memory_size))
+                if self.policy.use_recurrent
+                else None
+            )
 
         # Convert to tensors
+        current_obs = ObsUtil.from_buffer(batch, n_obs)
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:146]: Duplicate with line below?

**Review by andrewcoh — APPROVED** [review]

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py (hunk 4 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py]: Merged into 1 line

**Pull request merged**